crates/matrix-sdk/src/sliding_sync/cache.rs

```diff
@@ -64,16 +67,33 @@ pub(super) async fn store_sliding_sync_state(
     to_device_token: Option<String>,
 ) -> Result<()> {
     let storage_key = &sliding_sync.inner.storage_key;
+    let instance_storage_key = format_storage_key_for_sliding_sync(storage_key);
 
-    trace!(storage_key, "Saving a `SlidingSync`");
+    #[cfg(feature = "e2e-encryption")]
+    {
+        trace!(storage_key, "Saving a `SlidingSync`'s to-device token to the crypto store");
+        if let Some(olm_machine) = &*sliding_sync.inner.client.base_client().olm_machine().await {
```

Comment:
I don't think that this is the correct place to store the to-device token anymore. I think we should modify `OlmMachine::receive_sync_changes()` to accept a token as well. We should probably add a struct like `SyncChanges` or `EncryptionSyncChanges` which collects the now growing list of arguments `receive_sync_changes()` accepts.


This does two things:

1. Persisting the token happens atomically at the same time all the other sync changes get persisted.
2. It ensures that the old sync codepath will persist the to-device token in the same place as well, which should ease up things if the dedicated to-device endpoint becomes a reality.

Reply:
Makes sense. Quick question about the other user of `receive_sync_changes`, that's a FFI method wrapping it: should we expect that the caller of this method forwards us the to-device next_batch token too? As far as I understand, before the to-device endpoint becomes a reality that would just be the same as the `/sync` token (which is stored elsewhere) and it would be unused. So we could make it `Option`al there.

Reply:
Good question, I think we can make it non-optional. It actually won't be unused, it can now be used in the `/keys/query` requests the `OlmMachine` generates: https://github.com/matrix-org/matrix-rust-sdk/blob/020d5aa29201760aecac83065e1209c0f4a67a7d/crates/matrix-sdk-crypto/src/requests.rs#L188-L193.

Now that I link to this in the Rust SDK I wonder why the hell can't I find this parameter in the spec anymore.

In any case, I think storing the token twice, while I'm not anymore certain that it's always useful, won't hurt.

Reply:
Hmm, that token is unused in the code base, and the synapse implementation of that endpoint doesn't try to use a token. I'll remove this.